### ISSUE TITLE
Is payload text can save the detail exception or custom information from activity?

### ISSUE THREAD

**Issue by jarvan-jiang** [microsoft/durabletask-mssql#230]
Right now if a TaskFailedException is thrown in execution of Activity, the reason can be saved into payload.
But how to save the other information like exception detail into payload text too?

**Comment by davidmrdavid** [comment]
Hi @jarvan-jiang - can you please clarify your question? A concrete example of what you're looking to save would help me understand. Thanks

**Comment by jarvan-jiang** [comment]
> Hi @jarvan-jiang - can you please clarify your question? A concrete example of what you're looking to save would help me understand. Thanks

Hi @davidmrdavid , sorry make you confusion, let me try to explain more.
What I mean is, when I throw `TaskFailureException` in an Activity, the first parameter `reason` of `TaskFailureException` will be stored in `Payloads` table, which is nice. However, the rest like `innerException`  and `details` are not saved.
```
        public TaskFailureException(string reason, Exception innerException, string details)
            : base(reason, innerException)
        {
            Details = details;
        }
```
Is there is a way to save more info from the rest parameters than reason? Or I have to manually put all the info into reason?

**Comment by davidmrdavid** [comment]
Thanks @jarvan-jiang.

I did a bit triage here, the answer is not immediately obvious to me, but I wanted to document some findings here.

First - the `Payload` table schema is specified here:

https://github.com/microsoft/durabletask-mssql/blob/e5155826631f8abc8852c1adeb41e958ca4870d7/src/DurableTask.SqlServer/Scripts/schema-1.0.0.sql#L29-L41

There I can see a `Text` field that I expect may serve as a "catch all" to serialize all sorts of information. This your exception details would probably go, I think.

I also see that when we read from the database, we do try to read the failure details, at least if the if-statement branch in the snippet below is taken:

https://github.com/microsoft/durabletask-mssql/blame/e5155826631f8abc8852c1adeb41e958ca4870d7/src/DurableTask.SqlServer/SqlUtils.cs#L171-L187

and I see that if-statement branch is taken if the `Payload` column contains a JSON:

https://github.com/microsoft/durabletask-mssql/blob/e5155826631f8abc8852c1adeb41e958ca4870d7/src/DurableTask.SqlServer/SqlUtils.cs#L225-L237

This was implemented here: https://github.com/microsoft/durabletask-mssql/pull/188

So this brings me to a few questions:
(1) What package and version of DurableTask MSSQL are you using? Please list the package name (there's multiple, depending on whether you're running DTFx directly, or running Durable Functions / serverless) and the version?
(2) Do you see anything in the Payload column for your task failures? 
(3) Looking at the [aformentioned PR](https://github.com/microsoft/durabletask-mssql/pull/188), there should be a new configuration called `ErrorPropagationMode.UseFailureDetails` which should serialize these extra details. Is that something you've set or can try setting? If you can try setting it and reproduce with it, that would help us proceed further.

I also see a report [here](https://github.com/microsoft/durabletask-mssql/issues/209) that the new failure mode may have some edge cases when sub-orchestrators and their activities are involved. Please take a look and let me know if this matches your symptoms. Thanks!

**Comment by jarvan-jiang** [comment]
> Thanks [@jarvan-jiang](https://github.com/jarvan-jiang).
> 
> I did a bit triage here, the answer is not immediately obvious to me, but I wanted to document some findings here.
> 
> First - the `Payload` table schema is specified here:
> 
> [durabletask-mssql/src/DurableTask.SqlServer/Scripts/schema-1.0.0.sql](https://github.com/microsoft/durabletask-mssql/blob/e5155826631f8abc8852c1adeb41e958ca4870d7/src/DurableTask.SqlServer/Scripts/schema-1.0.0.sql#L29-L41)
> 
> Lines 29 to 41 in [e515582](/microsoft/durabletask-mssql/commit/e5155826631f8abc8852c1adeb41e958ca4870d7)
> 
>  IF OBJECT_ID(N'__SchemaNamePlaceholder__.Payloads', 'U') IS NULL 
>  BEGIN 
>      CREATE TABLE __SchemaNamePlaceholder__.Payloads ( 
>          [TaskHub] varchar(50) NOT NULL, 
>          [InstanceID] varchar(100) NOT NULL, 
>          [PayloadID] uniqueidentifier NOT NULL, 
>          [Text] varchar(max) NULL, 
>          [Reason] varchar(max) NULL, 
>          -- NOTE: no FK constraint to Instances table because we want to allow events to create new instances 
>   
>          CONSTRAINT PK_Payloads PRIMARY KEY (TaskHub, InstanceID, PayloadID), 
>      ) 
>  END 
> There I can see a `Text` field that I expect may serve as a "catch all" to serialize all sorts of information. This your exception details would probably go, I think.
> 
> I also see that when we read from the database, we do try to read the failure details, at least if the if-statement branch in the snippet below is taken:
> 
> https://github.com/microsoft/durabletask-mssql/blame/e5155826631f8abc8852c1adeb41e958ca4870d7/src/DurableTask.SqlServer/SqlUtils.cs#L171-L187
> 
> and I see that if-statement branch is taken if the `Payload` column contains a JSON:
> 
> [durabletask-mssql/src/DurableTask.SqlServer/SqlUtils.cs](https://github.com/microsoft/durabletask-mssql/blob/e5155826631f8abc8852c1adeb41e958ca4870d7/src/DurableTask.SqlServer/SqlUtils.cs#L225-L237)
> 
> Lines 225 to 237 in [e515582](/microsoft/durabletask-mssql/commit/e5155826631f8abc8852c1adeb41e958ca4870d7)
> 
>  static bool TryGetFailureDetails(DbDataReader reader, out FailureDetails? details) 
>  { 
>      // Failure details are expected to be in JSON format. In previous versions, it was just an error message 
>      // that isn't expected to be JSON. 
>      string? text = GetPayloadText(reader); 
>      if (string.IsNullOrEmpty(text) || text![0] != '{') 
>      { 
>          details = null; 
>          return false; 
>      } 
>   
>      return DTUtils.TryDeserializeFailureDetails(text, out details); 
>  } 
> This was implemented here: [#188](https://github.com/microsoft/durabletask-mssql/pull/188)
> 
> So this brings me to a few questions: (1) What package and version of DurableTask MSSQL are you using? Please list the package name (there's multiple, depending on whether you're running DTFx directly, or running Durable Functions / serverless) and the version? (2) Do you see anything in the Payload column for your task failures? (3) Looking at the [aformentioned PR](https://github.com/microsoft/durabletask-mssql/pull/188), there should be a new configuration called `ErrorPropagationMode.UseFailureDetails` which should serialize these extra details. Is that something you've set or can try setting? If you can try setting it and reproduce with it, that would help us proceed further.
> 
> I also see a report [here](https://github.com/microsoft/durabletask-mssql/issues/209) that the new failure mode may have some edge cases when sub-orchestrators and their activities are involved. Please take a look and let me know if this matches your symptoms. Thanks!

Thanks @davidmrdavid for your detail information. Recently I finally have time to follow your instruction to investigate that, by setting the `ErrorPropagationMode` you mentioned, there is another option need to be set is `IncludeDetails`, with this two options set to hub, I was able to log the exception detail into db and able to read it out. 
Thanks for your help!
